File: src/tq/rules/structure_mismatch.py
"""Structure rule for unit test placement mismatches."""

from __future__ import annotations

from pathlib import Path
from typing import TYPE_CHECKING

from tq.engine.models import Finding, Severity
from tq.engine.rule_id import RuleId

if TYPE_CHECKING:
    from tq.engine.context import AnalysisContext


class StructureMismatchRule:
    """Emit findings for test files that do not mirror source structure."""

    @property
    def rule_id(self) -> RuleId:
        """Return stable rule identifier."""
        return RuleId("structure-mismatch")

    def evaluate(self, context: AnalysisContext) -> tuple[Finding, ...]:
        """Evaluate test structure alignment against the analysis index."""
        package_path = _package_path_from_context(context)
        test_root_display = _test_root_display_from_context(context)
        known_target_paths = _known_target_paths_from_context(context)
        source_files = set(context.index.source_files)
        findings: list[Finding] = []

        for test_file in context.index.test_files:
            if _is_non_unit_test_path(test_file):
                continue

            if not _is_unit_test_filename(test_file.name):
                continue

            if not _starts_with_path_prefix(test_file=test_file, prefix=package_path):
                if _belongs_to_other_target(
                    test_file=test_file,
                    active_package_path=package_path,
                    known_target_paths=known_target_paths,
                ):
                    continue

                suggestion_path = (
                    test_root_display / package_path / test_file.name
                ).as_posix()

                findings.append(
                    Finding(
                        rule_id=self.rule_id,
                        severity=Severity.WARNING,
                        message="Unit test is not located under the package test root",
                        path=context.index.test_root / test_file,
                        suggestion=f"Move test under: {suggestion_path}",
                    ),
                )
                continue

            expected_path = _expected_path_for_test_file(
                test_file=test_file,
                source_files=source_files,
                package_path=package_path,
            )
            if expected_path is None or expected_path == test_file:
                continue

            findings.append(
                Finding(
                    rule_id=self.rule_id,
                    severity=Severity.WARNING,
                    message="Test file is not in the expected location",
                    path=context.index.test_root / test_file,
                    suggestion=(
                        f"Move to: {(test_root_display / expected_path).as_posix()}"
                    ),
                ),
            )

        return tuple(findings)


def _expected_path_for_test_file(
    *,
    test_file: Path,
    source_files: set[Path],
    package_path: Path,
) -> Path | None:
    """Infer canonical path when target source module can be resolved."""
    source_candidate = _resolve_source_candidate(
        test_file=test_file,
        source_files=source_files,
        package_path=package_path,
    )
    if source_candidate is None:
        return None

    expected_name = _expected_test_name(test_file=test_file)
    return package_path / source_candidate.parent / expected_name


def _resolve_source_candidate(
    *,
    test_file: Path,
    source_files: set[Path],
    package_path: Path,
) -> Path | None:
    """Resolve source file for a test path using deterministic candidates."""
    module_stem = test_file.stem[5:]
    candidates = _candidate_source_paths(
        test_file=test_file,
        module_stem=module_stem,
        package_path=package_path,
    )
    for candidate in candidates:
        if candidate in source_files:
            return candidate

    bare_name = f"{module_stem.split('_')[0]}.py"
    same_name_sources = [path for path in source_files if path.name == bare_name]
    if len(same_name_sources) == 1:
        return same_name_sources[0]

    return None


def _candidate_source_paths(
    *,
    test_file: Path,
    module_stem: str,
    package_path: Path,
) -> tuple[Path, ...]:
    """Build source candidates from unit test file names."""
    prefix_len = len(package_path.parts)
    relative_parts = test_file.parts[prefix_len:-1]
    direct_source = Path(*relative_parts) / f"{module_stem}.py"

    if "_" not in module_stem:
        return (direct_source,)

    prefixes: list[Path] = [direct_source]
    name_parts = module_stem.split("_")
    for index in range(len(name_parts) - 1, 0, -1):
        prefix = "_".join(name_parts[:index])
        prefixes.append(Path(*relative_parts) / f"{prefix}.py")

    return tuple(prefixes)


def _expected_test_name(*, test_file: Path) -> str:
    """Return canonical test file name preserving qualifiers."""
    return test_file.name


def _is_non_unit_test_path(test_file: Path) -> bool:
    """Check whether path is integration or end-to-end scope."""
    return "integration" in test_file.parts or "e2e" in test_file.parts


def _is_unit_test_filename(filename: str) -> bool:
    """Check if filename follows unit test naming shape."""
    return filename.startswith("test_") and filename.endswith(".py")


def _package_path_from_context(context: AnalysisContext) -> Path:
    """Resolve target package path from context settings."""
    value = context.settings.get("package_path")
    if isinstance(value, str) and value.strip():
        return Path(value)

    return Path(context.index.source_root.name)


def _test_root_display_from_context(context: AnalysisContext) -> Path:
    """Resolve display path prefix for test-root-relative suggestions."""
    value = context.settings.get("test_root_display")
    if isinstance(value, str) and value.strip():
        return Path(value)

    return Path(context.index.test_root.name)


def _known_target_paths_from_context(context: AnalysisContext) -> tuple[Path, ...]:
    """Resolve all configured target package roots from context settings."""
    values = context.settings.get("known_target_package_paths")
    if not isinstance(values, tuple):
        return ()

    return tuple(
        Path(value) for value in values if isinstance(value, str) and value.strip()
    )


def _starts_with_path_prefix(*, test_file: Path, prefix: Path) -> bool:
    """Check whether test file path starts with prefix parts."""
    prefix_parts = prefix.parts
    if len(test_file.parts) < len(prefix_parts):
        return False
    return test_file.parts[: len(prefix_parts)] == prefix_parts


def _belongs_to_other_target(
    *,
    test_file: Path,
    active_package_path: Path,
    known_target_paths: tuple[Path, ...],
) -> bool:
    """Check whether test file belongs to a sibling configured target."""
    for candidate_path in known_target_paths:
        if candidate_path == active_package_path:
            continue
        if _starts_with_path_prefix(test_file=test_file, prefix=candidate_path):
            return True
    return False

File: crates/tq-rules/src/structure_mismatch.rs
use std::collections::BTreeSet;
use std::path::{Path, PathBuf};

use tq_core::RelativePathBuf;
use tq_engine::{AnalysisContext, Finding, Rule, RuleId};

use crate::builtin::{
    BuiltinRule, is_non_unit_test_path, is_unit_test_filename, path_to_forward_slashes,
    starts_with_path_prefix,
};

pub struct StructureMismatchRule {
    rule_id: RuleId,
}

impl StructureMismatchRule {
    pub fn new() -> Result<Self, crate::error::RulesError> {
        Ok(Self {
            rule_id: BuiltinRule::StructureMismatch.rule_id()?,
        })
    }
}

impl Rule for StructureMismatchRule {
    fn rule_id(&self) -> &RuleId {
        &self.rule_id
    }

    fn evaluate(&self, context: &AnalysisContext) -> Vec<Finding> {
        let package_path = context.package_path();
        let test_root_display = context.test_root_display();
        let known_target_paths = context.known_target_package_paths();
        let source_files = context
            .index()
            .source_files()
            .iter()
            .cloned()
            .collect::<BTreeSet<_>>();

        let mut findings = Vec::new();
        for test_file in context.index().test_files() {
            if is_non_unit_test_path(test_file) {
                continue;
            }

            let Some(file_name) = test_file.file_name().and_then(std::ffi::OsStr::to_str) else {
                continue;
            };
            if !is_unit_test_filename(file_name) {
                continue;
            }

            if !starts_with_path_prefix(test_file, package_path) {
                if belongs_to_other_target(test_file, package_path, known_target_paths) {
                    continue;
                }

                let suggestion_path = test_root_display.join(package_path).join(file_name);
                if let Ok(finding) = Finding::new(
                    self.rule_id.clone(),
                    BuiltinRule::StructureMismatch.default_severity(),
                    "Unit test is not located under the package test root",
                    context.index().test_root().join(test_file),
                    None,
                    Some(format!(
                        "Move test under: {}",
                        path_to_forward_slashes(&suggestion_path)
                    )),
                    None,
                ) {
                    findings.push(finding);
                }
                continue;
            }

            let Some(expected_path) =
                expected_path_for_test_file(test_file, &source_files, package_path)
            else {
                continue;
            };
            if expected_path == *test_file {
                continue;
            }

            if let Ok(finding) = Finding::new(
                self.rule_id.clone(),
                BuiltinRule::StructureMismatch.default_severity(),
                "Test file is not in the expected location",
                context.index().test_root().join(test_file),
                None,
                Some(format!(
                    "Move to: {}",
                    path_to_forward_slashes(&test_root_display.join(expected_path))
                )),
                None,
            ) {
                findings.push(finding);
            }
        }

        findings
    }
}

fn expected_path_for_test_file(
    test_file: &Path,
    source_files: &BTreeSet<PathBuf>,
    package_path: &Path,
) -> Option<PathBuf> {
    let source_candidate = resolve_source_candidate(test_file, source_files, package_path)?;
    let expected_name = test_file.file_name()?;
    Some(
        package_path
            .join(source_candidate.parent()?)
            .join(expected_name),
    )
}

fn resolve_source_candidate(
    test_file: &Path,
    source_files: &BTreeSet<PathBuf>,
    package_path: &Path,
) -> Option<PathBuf> {
    let module_stem = test_file
        .file_stem()
        .and_then(std::ffi::OsStr::to_str)
        .and_then(|stem| stem.strip_prefix("test_"))?;
    for candidate in candidate_source_paths(test_file, module_stem, package_path) {
        if source_files.contains(&candidate) {
            return Some(candidate);
        }
    }

    let bare_name = format!("{}.py", module_stem.split('_').next().unwrap_or_default());
    let same_name_sources = source_files
        .iter()
        .filter(|path| {
            path.file_name()
                .is_some_and(|name| name == std::ffi::OsStr::new(&bare_name))
        })
        .collect::<Vec<_>>();
    if same_name_sources.len() == 1 {
        return same_name_sources.first().map(|path| (*path).clone());
    }

    None
}

fn candidate_source_paths(
    test_file: &Path,
    module_stem: &str,
    package_path: &Path,
) -> Vec<PathBuf> {
    let prefix_len = package_path.components().count();
    let test_components = test_file.components().collect::<Vec<_>>();
    let relative_components = test_components
        .iter()
        .skip(prefix_len)
        .take(test_components.len().saturating_sub(prefix_len + 1))
        .copied()
        .collect::<Vec<_>>();

    let relative_source_dir = relative_components
        .iter()
        .fold(PathBuf::new(), |path, component| {
            path.join(component.as_os_str())
        });

    let direct_source = relative_source_dir.join(format!("{module_stem}.py"));
    if !module_stem.contains('_') {
        return vec![direct_source];
    }

    let mut candidates = vec![direct_source];
    let parts = module_stem.split('_').collect::<Vec<_>>();
    for index in (1..parts.len()).rev() {
        let prefix = parts[..index].join("_");
        candidates.push(relative_source_dir.join(format!("{prefix}.py")));
    }

    candidates
}

fn belongs_to_other_target(
    test_file: &Path,
    active_package_path: &Path,
    known_target_paths: &[RelativePathBuf],
) -> bool {
    known_target_paths.iter().any(|path| {
        let known_path = path.as_path();
        known_path != active_package_path && starts_with_path_prefix(test_file, known_path)
    })
}
